File: solutions/swift/kindergarten-garden/1/Sources/KindergartenGarden/KindergartenGarden.swift
enum Plant: Character {
    case grass = "G"
    case clover = "C"
    case radishes = "R"
    case violets = "V"
}

struct Garden {
    private let rows: [String]

    private static let children = [
        "Alice", "Bob", "Charlie", "David",
        "Eve", "Fred", "Ginny", "Harriet",
        "Ileana", "Joseph", "Kincaid", "Larry"
    ]

    init(_ diagram: String) {
        rows = diagram.split(separator: "\n").map(String.init)
    }

    func plantsForChild(_ child: String) -> [Plant] {
        guard let childIndex = Garden.children.firstIndex(of: child) else {
            return []
        }

        return Garden.plants(from: rows, in: childIndex)
    }

    private static func plants(from rows: [String], in childIndex: Int) -> [Plant] {
        rows.flatMap { row in
            let i = row.index(row.startIndex, offsetBy: childIndex * 2)
            let j = row.index(after: i)

            return [row[i], row[j]].compactMap { Plant(rawValue: $0) }
        }
    }
}

File: solutions/euphoria/kindergarten-garden/1/kindergarten-garden.ex
function decode(integer plant)
    if plant = 'G' then
        return "grass"
    elsif plant = 'C' then
        return "clover"
    elsif plant = 'R' then
        return "radishes"
    elsif plant = 'V' then
        return "violets"
    end if

    return ""
end function

public function plants(sequence garden, sequence student)
    sequence students
    integer newline_pos
    sequence row1
    sequence row2
    integer student_pos
    integer cup_pos
    sequence result

    students = {
        "Alice", "Bob", "Charlie", "David",
        "Eve", "Fred", "Ginny", "Harriet",
        "Ileana", "Joseph", "Kincaid", "Larry"
    }

    newline_pos = find('\n', garden)
    row1 = garden[1..newline_pos-1]
    row2 = garden[newline_pos+1..length(garden)]

    student_pos = find(student, students)

    if student_pos = 0 then
        return {}
    end if

    cup_pos = (student_pos - 1) * 2 + 1

    result = {
        decode(row1[cup_pos]),
        decode(row1[cup_pos + 1]),
        decode(row2[cup_pos]),
        decode(row2[cup_pos + 1])
    }

    return result
end function